cpp/src/fil/batch_tree_reorg.cu

```diff
@@ -0,0 +1,126 @@
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include "common.cuh"
+
+namespace ML {
+namespace fil {
+
+using namespace MLCommon;
+
+// vec wraps float[N] for cub::BlockReduce
+template <int N>
+struct vec {
+  float data[N];
+  __host__ __device__ float& operator[](int i) { return data[i]; }
+  __host__ __device__ float operator[](int i) const { return data[i]; }
+  friend __host__ __device__ vec<N> operator+(const vec<N>& a,
+                                              const vec<N>& b) {
+    vec<N> r;
+#pragma unroll
+    for (int i = 0; i < N; ++i) r[i] = a[i] + b[i];
+    return r;
+  }
+};
+
+template <int NITEMS>
+__device__ __forceinline__ void infer_one_tree(const dense_node* root, int tree,
+                                               float* sdata, int depth,
+                                               int ntrees, int cols,
+                                               vec<NITEMS>& out) {
+  int curr[NITEMS];
+  int mask = (1 << NITEMS) - 1;  // all active
+  for (int j = 0; j < NITEMS; ++j) curr[j] = 0;
+  do {
+#pragma unroll
+    for (int j = 0; j < NITEMS; ++j) {
+      if ((mask >> j) & 1 == 0) continue;
+      dense_node n = root[curr[j] * ntrees + tree];
+      if (n.is_leaf()) {
+        mask &= ~(1 << j);
+        continue;
+      }
+      float val = sdata[j * cols + n.fid()];
+      bool cond = isnan(val) ? !n.def_left() : val >= n.thresh();
+      curr[j] = (curr[j] << 1) + 1 + cond;
+    }
+  } while (mask != 0);
+  for (int j = 0; j < NITEMS; ++j)
+    out[j] += root[curr[j] * ntrees + tree].output();
+}
+
+template <int NITEMS>
+__global__ void batch_tree_reorg_kernel(predict_params ps) {
+  // cache the row for all threads to reuse
+  extern __shared__ char smem[];
+  float* sdata = (float*)smem;
+  size_t rid = blockIdx.x * NITEMS;
+  for (int j = 0; j < NITEMS; ++j) {
+    for (int i = threadIdx.x; i < ps.cols; i += blockDim.x) {
+      size_t row = rid + j;
+      sdata[j * ps.cols + i] =
+        row < ps.rows ? ps.data[row * ps.cols + i] : 0.0f;
+    }
+  }
+  __syncthreads();
+  // one block works on a single row and the whole forest
+  vec<NITEMS> out;
+  for (int i = 0; i < NITEMS; ++i) out[i] = 0.0f;
+  int max_nodes = tree_num_nodes(ps.depth);
+  for (int j = threadIdx.x; j < ps.ntrees; j += blockDim.x) {
+    infer_one_tree<NITEMS>(ps.nodes, j, sdata, ps.depth, ps.ntrees, ps.cols,
+                           out);
+  }
+  typedef cub::BlockReduce<vec<NITEMS>, TPB> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage tmp_storage;
+  out = BlockReduce(tmp_storage).Sum(out);
+  if (threadIdx.x == 0) {
+    for (int i = 0; i < NITEMS; ++i) {
+      int idx = blockIdx.x * NITEMS + i;
+      if (idx < ps.rows) ps.preds[idx] = out[i];
+    }
+  }
+}
+
+const static int MAX_NITEMS = 4;
+void batch_tree_reorg(const predict_params& ps, cudaStream_t stream) {
+  int nitems = ps.max_shm / (sizeof(float) * ps.cols);
+  ASSERT(nitems > 0, "p.cols == %d: too many columns", ps.cols);
```

Comment:
Should we also print maximum supported column here? Also is 'columns' a common way to refer to attributes/features of samples?